rosbag2/include/rosbag2/compression_interfaces/base_compressor_interface.hpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2__COMPRESSION_INTERFACES__BASE_COMPRESSOR_INTERFACE_HPP_
+#define ROSBAG2__COMPRESSION_INTERFACES__BASE_COMPRESSOR_INTERFACE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rosbag2/visibility_control.hpp"
+
+namespace rosbag2
+{
+class ROSBAG2_PUBLIC BaseCompressorInterface
+{
+public:
+  virtual ~BaseCompressorInterface() {}
+
+  /**
+   * Compress a file on disk.
+   *
+   * \param uri Input file to compress with file extension.
+   * \return The relative path to the compressed file with the compressed extension.
+   */
+  virtual std::string compress_uri(const std::string & uri) = 0;
+
+  /**
+   * Compress serialized_data in a serialized bag message.
+   *
+   * @param bag_message A serialized bag message.
+   * @return A shared pointer to the bag message with compressed serialized_data.
+   */
+  virtual std::shared_ptr<SerializedBagMessage> compress_serialized_bag_message(
```

Comment:
Is it intentional that `bag_messages` is passed by reference?
Elsewhere, `std::shared_ptr<SerializedBagMessage>` is passed by value. If we're going to go against that design choice, I think we should instead pass as a raw pointer.

Reply:
Lets stick to convention.

Reply:
Shouldn't that bag message be passed in `const`? I think it'd be a bit strange if the original message was modified. 

Also, does it have to be a `SerializedBagMessage`? I could see that a serialized message would suffice and so this functionality could even live outside of rosbag2 completely.

Reply:
how was that resolved?